### ISSUE TITLE
GADTs are ignored in the newtype hint

### ISSUE THREAD

**Issue by googleson78** [ndmitchell/hlint#697]
`ghc` accepts both
```haskell
data X where
  X :: Int -> X
```
and
```haskell
newtype X where
  X :: Int -> X
```

but `hlint` doesn't suggest using a `newtype` in this case. This seems easily solvable after/during #645.

**Comment by ndmitchell** [comment]
Good catch - worth fixing. Happy for it to happen simultaneously with the refactor or after. One thing to note is that an existential in a GADT doesn't have to use forall, and existentials can't be converted to newtype.

**Comment by googleson78** [comment]
Okay, I'll do it with the rest of #696

I think that `con_qvars` from [`ConDeclGADT`](https://hackage.haskell.org/package/ghc-lib-parser-8.8.0.20190424/docs/HsDecls.html#t:ConDecl) will do the job here.

**Comment by googleson78** [comment]
Note:
I decided to not do it with the rest of #696, becuase it turned out to actually require *a bit* of code. This is due to the fact that a newtype GADT constructor needs to always return the same type as the type being declared. It also seems quite low-priority due to this limitation (only GADT syntax aficionados seem likely to benefit from this.)

**Comment by googleson78** [comment]
This turns out to be way more annoying than expected, because of [Note [GADT abstract syntax]](https://hackage.haskell.org/package/ghc-8.10.1/docs/src/GHC.Hs.Decls.html#HsConDeclDetails).
Basically we don't actually have access to the arguments of a GADT constructor during the parsing phase (they're filled in during renaming)

In order to implement this there would need to be some heuristic on matching `HsType`s for "simple GADT constructors".

So things like `a -> b` where
* `a` is a concrete type, a type containing only bound variables etc (not sure myself exactly what this should be)
* `b` is `Foo` where `Foo` is the name of the gadt + all its type variables "intact" (none of them are instantiated to anything, because then a newtype GADT isn't allowed)

As a free bonus, implicit qualified type vars (so things like the `b` in `data Foo where b -> Foo`) aren't filled in before renaming either, so this heuristic needs to detect type vars too.

**Comment by ndmitchell** [comment]
Thanks for investigating. That is grim! My inclination is that the cost of implementing such a hint is probably greater than the value it brings, and I especially worry about detecting type vars and concrete types etc.

**Comment by ndmitchell** [comment]
Hmm, reading the note in more detail, if we only worked with types with no symbols in them (other than `->`) we probably get most of the cases without a lot of the problems. Still unsure if it's worth it, but looks a bit more feasible.
